### ISSUE TITLE
Edges shifted and mirrored in edge images after batch_hed.py and PostprocessHED.m

### ISSUE THREAD

**Issue by diaosiji** [junyanz/pytorch-CycleGAN-and-pix2pix#911]
I am **reproducing the edges2shoes dataset for learning how to extract edges from image object by using HED**, it almost success expect a little strange problem: **part of the edges shifed and mirrored**. It looks like below, they are different from the pix2pix paper.

![1_AB](https://user-images.githubusercontent.com/11862759/73901207-1077be80-48cd-11ea-96c2-1ad02f7b8b56.jpg) ![1_AB1](https://user-images.githubusercontent.com/11862759/73901241-22f1f800-48cd-11ea-977a-8a9777ef2320.jpg)

![7_AB](https://user-images.githubusercontent.com/11862759/73901264-30a77d80-48cd-11ea-80a7-e67b79bc43c6.jpg) ![7_AB1](https://user-images.githubusercontent.com/11862759/73901294-52086980-48cd-11ea-8b47-4b4ec78a5093.jpg)

**My process followed the instruction  strictly, which is at the top annotation of batch_hed.py and PostprocessHED.m**. such as in batch_hed.py:

 HED batch processing script; modified from https://github.com/s9xie/hed/blob/master/examples/hed/HED-tutorial.ipynb
Step 1: download the hed repo: https://github.com/s9xie/hed
Step 2: download the models and protoxt, and put them under {caffe_root}/examples/hed/
Step 3: put this script under {caffe_root}/examples/hed/
Step 4: run the following script: python batch_hed.py --images_dir=/data/to/path/photos/ --hed_mat_dir=/data/to/path/hed_mat_files/
The code sometimes crashes after computation is done. Error looks like "Check failed: ... driver shutting down". You can just kill the job.
For large images, it will produce gpu memory issue. Therefore, you better resize the images before running this script.
Step 5: run the MATLAB post-processing script "PostprocessHED.m"

and in PostprocessHED.m:

 HED batch processing script; modified from https://github.com/s9xie/hed/blob/master/examples/hed/HED-tutorial.ipynb
 Step 1: download the hed repo: https://github.com/s9xie/hed
 Step 2: download the models and protoxt, and put them under {caffe_root}/examples/hed/
 Step 3: put this script under {caffe_root}/examples/hed/
 Step 4: run the following script: python batch_hed.py --images_dir=/data/to/path/photos/ --hed_mat_dir=/data/to/path/hed_mat_files/
The code sometimes crashes after computation is done. Error looks like "Check failed: ... driver shutting down". You can just kill the job.
For large images, it will produce gpu memory issue. Therefore, you better resize the images before running this script.
Step 5: run the MATLAB post-processing script "PostprocessHED.m"

Some facts may help:
- I **used a offical caffe docker** for convenience;
- The images are form the edges2shoes dataset, and i cut the paired image in two: one edge image and one object;
- image size is 256x256;
- In PostprocessHED.m it error frist, said can't find 'predict' variable, I changed the code in function GetEdge, from load(filePath) to load(filePath, 'predict');

I got stuck here for a long time, searched many issues and can not find a similar one, so I opened a new issue for it. Could you please take a look at this issue ? Thanks a lot ! @junyanz

**Comment by junyanz** [comment]
I am not sure what happened. Maybe you want to check the flag `--border`. Here is the [line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/scripts/edges/batch_hed.py#L28)

**Comment by diaosiji** [comment]
> I am not sure what happened. Maybe you want to check the flag `--border`. Here is the [line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/scripts/edges/batch_hed.py#L28)

Thanks, I analyze some factors, which including it. I will try it today.

**Comment by diaosiji** [comment]
<img width="1063" alt="截屏2020-02-07下午4 18 38" src="https://user-images.githubusercontent.com/11862759/74013292-105be980-49c7-11ea-8fac-2a9e0bf51de9.png">
I **tried several border value**, from 128 to 8 (128, 64, 32, 16, 8), it seems that border is not the answer.

**Comment by diaosiji** [comment]
Continue. In order to **locate problem**, I add some code at the end of batch_hed.py, to **visualize the fuse ndarray, which is the output of hed model**. The codes are as follow:
` 
  
    #  test code for fuse visualization
    # get to know fuse data structure 
    #print(type(fuse))
    #print(fuse.size)
    #print(fuse.shape)
    # clip float32 value range to [0.0-1.0]
    fuse = np.clip(fuse, 0.0, 1.0)
    # output image
    Image.fromarray(((fuse - 1.0) * (- 255.0)).astype(np.uint8)).save('./out.png')
`
Then I got the image, it shows that the part of the edges shifed and mirrored problem is still here, however, now I know the problem is in batch_hed.py. The image is bellow:
![out](https://user-images.githubusercontent.com/11862759/74094959-7ca02f80-4b24-11ea-93a6-7a3254c2a1a9.png)

**Comment by diaosiji** [comment]
Continue. Go on locating the problem, I annotate the border cuting code to see the difference of images:

`# get rid of the border`
`#fuse = fuse[border:-border, border:-border]`

and the result is interesting, I got a 512*512 image, and it looks like:

=================================

![2](https://user-images.githubusercontent.com/11862759/74205589-fe25c800-4cb2-11ea-944b-44ea707546bd.png)

================================= 

So, I realized that **there must some code let the whole edges shift right and down**, I can get the right edges!

**Comment by diaosiji** [comment]
I used GIMP to measure the shift pix, it seems 32 pix. hum......

**Comment by diaosiji** [comment]
Find an simple and crude solution: 

`# get rid of the border`
`#fuse = fuse[border:-border, border:-border]`
`# from 32+128 to 32+128+256,the 32 can be adjust! 34,35......`
`fuse = fuse[163:419, 163:419]`

And the result is below:

=============================

![3](https://user-images.githubusercontent.com/11862759/74216412-52df3800-4cdf-11ea-837a-c6e466b0f9b2.png)


=============================

However, this solution did not get the reason why that shift happen, so I will continue.

**Comment by diaosiji** [comment]
Finally, after days of tries and errors, I am sure that the shift comes from the pretrained model, since I visualized every line of codes before the net.forward(), and there is no shift. So I suppose that it is the pretrained model which cause the shift.

`

    in_ = np.array(im, dtype=np.float32)
    # There are 11 mode:constant,reflect,edge and so on
    in_ = np.pad(in_, ((border, border), (border, border), (0, 0)), 'reflect')

    in_ = in_[:, :, 0:3]
    # switch to BGR
    in_ = in_[:, :, ::-1]
    # subtract mean
    in_ -= np.array((104.00698793, 116.66876762, 122.67891434))
    # make dims C x H x W for Caffe
    in_ = in_.transpose((2, 0, 1))

    # remove the following two lines if testing with cpu

    # shape for input (data blob is N x C x H x W), set data
    net.blobs['data'].reshape(1, *in_.shape)
    net.blobs['data'].data[...] = in_
    # run net and take argmax for prediction
    net.forward()
    fuse = net.blobs['sigmoid-fuse'].data[0][0, :, :]
`

================================================================

![out](https://user-images.githubusercontent.com/11862759/74516320-2891b280-4f4b-11ea-948c-1c0abbf40de7.png) 

![test2](https://user-images.githubusercontent.com/11862759/74516353-35aea180-4f4b-11ea-90c6-2074fd7781f4.png) 

![test3](https://user-images.githubusercontent.com/11862759/74516380-3d6e4600-4f4b-11ea-9c15-00579bf047df.png)

=================================================================

Done, with a little upset.

**Comment by diaosiji** [comment]
Thank you @junyanz !

**Comment by junyanz** [comment]
Great observation. Thank you for reporting it.
